controller/tests/controller_server/utils_test.py

```diff
@@ -22,32 +23,116 @@ def setUp(self):
+    def test_validate_secrets_with_no_username(self):
+        secrets = {"password": password, "management_address": array}
+        self._test_validate_secrets_validation_errors(secrets)
 
+    def test_validate_secrets_with_empty_dict(self):
         secrets = {}
-        with self.assertRaises(ValidationException):
-            utils.validate_secret(secrets)
+        self._test_validate_secrets_validation_errors(secrets)
+
+    def test_validate_secrets_with_config(self):
+        secrets = {"config": str({"u1": {"username": user, "password": password, "management_address": array}}
+                                 )}
+        utils.validate_secrets(secrets)
+
+    def test_validate_secrets_with_config_invalid_secret(self):
+        secrets = {"config": str({"u1": {"username": user, "management_address": array}})}
+        self._test_validate_secrets_validation_errors(secrets)
+
+    def _test_validate_secrets_with_config_valid_uid(self, uid):
+        secrets = {"config": str({uid: {"username": user, "password": password, "management_address": array}}
+                                 )}
+        utils.validate_secrets(secrets)
+
+    def test_validate_secrets_with_config_valid_uid(self):
+        self._test_validate_secrets_with_config_valid_uid("u i_d.")
+        self._test_validate_secrets_with_config_valid_uid("a" * 90)
+
+    def test_validate_secrets_with_config_and_topologies(self):
+        secrets = {"config": str({"u1": {"username": user, "password": password, "management_address": array,
+                                         "supported_topologies": [{"topology.kubernetes.io/test": "zone1",
+                                                                   "topology.block.csi.ibm.com/test": "dev1"}]}}
+                                 )}
+        utils.validate_secrets(secrets)
+
+    def _test_validate_secrets_with_config_invalid_parameters(self, uid="u1", topology="topology.kubernetes.io/test",
+                                                              value="test"):
+        secrets = {"config": str({uid: {"username": user, "password": password, "management_address": array,
+                                        "supported_topologies": [{topology: value}]}})}
+        self._test_validate_secrets_validation_errors(secrets)
+
+    def test_validate_secrets_with_config_invalid_parameters(self):
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u-1")
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u:1")
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u1+")
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u1*")
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u-1(")
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u/1")
+        self._test_validate_secrets_with_config_invalid_parameters(uid="u=1")
+        self._test_validate_secrets_with_config_invalid_parameters(uid=" ")
```

Comment:
I suggest adding an empty string test and a None test if it can theoretically happen.

Reply:
nice, Im not sure that None is possible but added another check for both